### PR TITLE
Replace resource mapper with dynamicRESTMapper

### PR DESCRIPTION
**Pull request by elgnay**
The related issue: https://github.com/open-cluster-management/backlog/issues/8944

Signed-off-by: Yang Le <yangle@redhat.com>

### CODE REVIEW

File: pkg/spoke/resource/mapper.go [hunk 1]
```diff
@@ -35,19 +36,30 @@ func (p *Mapper) Run(stopCh <-chan struct{}) {
 	go wait.Until(func() {
 		p.syncLock.Lock()
 		defer p.syncLock.Unlock()
-		deferredMappd := p.Mapper.(*restmapper.DeferredDiscoveryRESTMapper)
+		deferredMappd := p.StandbyMapper.(*restmapper.DeferredDiscoveryRESTMapper)
 		deferredMappd.Reset()
+		// swap two mapper if the standby mapper is reset and ready to serve
+		if ready(p.StandbyMapper) {
+			p.ServingMapper, p.StandbyMapper = p.StandbyMapper, p.ServingMapper
```
Comment (qiujian16) [inline pkg/spoke/resource/mapper.go]: I think we could put lock within this if?

File: pkg/spoke/resource/mapper.go [hunk 2]
```diff
@@ -35,19 +36,30 @@ func (p *Mapper) Run(stopCh <-chan struct{}) {
...
 	if err != nil {
 		return nil, fmt.Errorf("the server doesn't have a resource type %q", gvk.Kind)
 	}
 
 	return mapping, nil
 }
+
+// ready returns true if the given mapper is ready to serve
+func ready(mapper meta.RESTMapper) bool {
+	// if the mapper is able to handle at lease 1 group kind it is ready to serve
```
Comment (qiujian16) [inline pkg/spoke/resource/mapper.go]: at least

**Review by qiujian16 — COMMENTED** [review]
i am also thinking whether NewDynamicRESTMapper is a better way to resolve our problem

File: pkg/spoke/resource/mapper.go (hunk 1 above)
Comment (elgnay) [reply pkg/spoke/resource/mapper.go]: Yes, you are right. We can do it for only `p.ServingMapper` might be accessed in different Goroutines.

File: pkg/spoke/resource/mapper.go (hunk 2 above)
Comment (elgnay) [reply pkg/spoke/resource/mapper.go]: Corrected.

**Comment by elgnay** [comment]
> i am also thinking whether NewDynamicRESTMapper is a better way to resolve our problem

I tried dynamicRESTMapper and it's good in our use case. It caches groupResources and try to refresh when the query does not hit the cache. I've refactored the code to use dynamicRESTMapper.

**Pull request merged**